docs/codeql/ql-language-reference/formulas.rst

```diff
@@ -399,7 +400,10 @@ With the following definition, an integer is in the class ``OneTwoThree`` if it
 ===========
 
 You can use the keyword ``implies`` between two formulas. The resulting formula is called an 
-implication. This is just a simplified notation: ``A implies B`` is the same as writing ``(not A) or B``.
+implication. This is just a simplified notation:
+
+* ``A implies B`` is the same as writing ``(not A) or B``.
+* ``A implies B`` is the same as writing ``if A then B else any()``.
```

Comment:
This is not quite right. It means the same, but is not operationally equivalent.

The desugaring for `if A then B else C` is in fact `(A and B) or ((not A) and C)`. Thus, if `C` is `any()`, the if/then/else desugars to `(A and B) or (not A)`. We can distribute the terms to see this is equivalent to `(A or (not A)) and (B or (not A))`, which simplifies to `(not A) or B`, so logically they are the same.

Practically, however, it may be that `(A and B)` is much faster to evaluate than just `B` -- perhaps `A` has very few tuples, while `B` has very many, and so it's faster to just check whether `B` holds of the tuples in `A` rather than enumerating all the tuples in `B`. Or, in other words, writing `if A then B else any()` evaluates the `then` part only for contexts that satisfy the condition, while this constraint does not exist in `A implies B`.

[You might argue that we should therefore desugar `A implies B` to `(not A) or (A and B)`, but that can be worse if `A` in fact does not significantly restrict the values in `B`, and therefore it would be a potentially breaking change. Still, it might be less surprising. 🤔]

```suggestion
* ``A implies B`` means the same as writing ``if A then B else any()``, though the latter is in fact evaluated as ``A implies (A and B)``.
```

Reply:
... I'm actually not sure whether we want this kind of operational detail in the language reference -- how do others feel?